docs/orchestration.asciidoc

```diff
@@ -0,0 +1,104 @@
+- rolling upgrades are performed in a safe way, reusing PersistentVolumes of the upgraded Elasticsearch nodes
+
+[id="{p}-statefulsets"]
+==== StatefulSets orchestration
+
+Behind the scenes, ECK translates each NodeSet specified in the Elasticsearch resource into a link:https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/[StatefulSet in Kubernetes]. The StatefulSet specification is based on the NodeSet specification:
+
+* `count` corresponds to the number of replicas in the StatefulSet, each replica leading to the creation of a Pod, which corresponds to a single Elasticsearch node
+* `podTemplate` is patched with additional required settings to run an Elasticsearch Pod, including the Elasticsearch container and sane default settings for the Pod
+* the StatefulSet name is built from the Elasticsearch resource name and the NodeSet name (`clustername-nodesetname`). Each Pod will be assigned the StatefulSet name, suffixed by an ordinal (`clustername-nodesetname-0`). The corresponding Elasticsearch node has the same name.
+
+The actual Pod creation is handled by the StatefulSet controller in Kubernetes. However, ECK sets the link:https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#on-delete[StatefulSet update strategy] to `OnDelete`, since it needs full control over when and how Pods get upgraded to a new revision.
+
+Whenever a Pod is removed and recreated (maybe with a newer revision), the StatefulSet controller ensures PersistentVolumes that were attached to the original Pod are attached again to the replacing Pod.
+
+[id="{p}-upgrade-patterns"]
+==== Cluster upgrade patterns
+
+Depending on how the NodeSets specification are updated, ECK handles the Kubernetes resources reconciliation in various ways.
+
+* When a new NodeSet is added to the Elasticsearch resource, ECK creates the corresponding StatefulSet. It also sets up link:https://kubernetes.io/docs/concepts/configuration/secret/[Secrets] and link:https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/[ConfigMaps] to hold the Elasticsearch configuration files and TLS certificates.
+* When the node count of an existing NodeSet is increased, ECK increases the replicas of the corresponding StatefulSet.
+* When the node count of an existing NodeSet is decreased, ECK migrates data away from the corresponding Elasticsearch nodes to remove, then decreases the replicas of the corresponding StatefulSet, once data migration is over. Corresponding <<{p}-volume-claim-templates,PersistentVolumeClaims are automatically removed>>.
+* When an existing NodeSet is removed, ECK migrates data away from the corresponding Elasticsearch nodes to remove, decreases the StatefulSet replicas accordingly, then finally removes the corresponding StatefulSet.
```

Comment:
It might be worth clarifying if we do it one node at a time or all at once here

Reply:
Can be both, depending on the updateStrategy.changeBudget. Which has its own section in the doc.